File: Plugins/Kingbotk/AWB Plugin/MyTrace.vb
Namespace AutoWikiBrowser.Plugins.SDKSoftware.Kingbotk.Components
    ''' <summary>
    ''' Logging manager
    ''' </summary>
    ''' <remarks></remarks>
    Friend NotInheritable Class MyTrace
        Inherits TraceManager

        Private WithEvents LoggingSettings As PluginLogging
        Private ActivePlugins As List(Of PluginBase)
        Private Shared LogFolder As String
        Private webcontrol As WikiFunctions.Browser.WebControl
        Private LoggedIn As Boolean, LoggedInUser As String
        Private BusyCounter As Integer = 0

        Private Const conBadPages As String = "Bad"
        Private Const conWiki As String = "Wiki"
        Private Const conXHTML As String = "XHTML"

        ' The most important stuff:
        Private Sub Trace_Upload(ByVal Sender As TraceListenerUploadableBase, ByRef Success As Boolean)
            Dim frm As New UploadingPleaseWaitForm
            Dim Uploader As New Uploader.LogUploader

            LoggingSettings.Led1.Colour = Colour.Blue
            Application.DoEvents()
            frm.Show()

            If Sender.TraceStatus.LogName = conBadPages Then
                ' TODO: BadPages upload
            Else
                With LoggingSettings.Settings
                    Dim UploadTo As String = .GlobbedUploadLocation & "/" & Sender.PageName.Replace( _
                        PluginLogging.Props.conUploadCategoryIsJobName, .Category)

                    Try
                        Uploader.LogIn(LoggingSettings.LoginDetails)
                        Application.DoEvents()

                        Uploader.LogIt(Sender, .LogTitle, .WikifiedCategory, UploadTo, .LinksToLog(ActivePlugins), _
                        .UploadOpenInBrowser, .UploadAddToWatchlist, PluginManager.UserName, _
                        "*" & PluginManager.conWikiPlugin & " version " & PluginLogging.Props.PluginVersion & _
                        Microsoft.VisualBasic.vbCrLf & "*[[WP:AWB|AWB]] version " & _
                        PluginLogging.Props.AWBVersion & Microsoft.VisualBasic.vbCrLf, _
                        PluginManager.conWikiPluginBrackets & " " & LogUploader.conUploadingDefaultEditSummary, _
                        PluginManager.conWikiPluginBrackets & " " & LogUploader.conAddingLogEntryDefaultEditSummary)

                        Success = True
                        DirectCast(Sender.TraceStatus, TraceStatus).UploadsCount += 1
                    Catch
                        Success = False
                    Finally
                        UploadTo += " " & Sender.TraceStatus.PageNumber.ToString
                        If Success Then
                            Sender.WriteCommentAndNewLine("Log uploaded to " & UploadTo)
                        Else
                            Sender.WriteCommentAndNewLine( _
                               "LOG UPLOADING FAILED. Please manually upload this section to " & UploadTo)
                        End If
                    End Try
                End With
            End If

            frm.Dispose()
            If BusyCounter = 0 Then LoggingSettings.Led1.Colour = Colour.Red _
               Else LoggingSettings.Led1.Colour = Colour.Green
        End Sub
        Friend Sub Initialise()
            LoggingSettings.Initialised = True
            LogFolder = LoggingSettings.Settings.LogFolder

            With LoggingSettings.Settings
                If .LogBadPages OrElse .LogXHTML OrElse .LogWiki Then
                    If Not IO.Directory.Exists(.LogFolder) Then LogFolder = Application.StartupPath

                    If .LogBadPages Then NewBadPagesTraceListener()
                    If .LogXHTML Then NewXHTMLTraceListener()
                    If .LogWiki Then NewWikiTraceListener()
                End If
                For Each t As KeyValuePair(Of String, IMyTraceListener) In Listeners
                    t.Value.WriteBulletedLine("Start button clicked. Initialising log.", True, False, True)
                Next
            End With
            CheckWeHaveLogInDetails()
        End Sub

        ' State:
        Friend ReadOnly Property HaveOpenFile() As Boolean
            Get
                Return Listeners.Count > 0
            End Get
        End Property

        ' Private:
        Private Sub CheckWeHaveLogInDetails()
            If Me.Uploadable AndAlso Not LoggingSettings.LoginDetails.IsSet Then
                LoggingSettings.LoginDetails = New LoginForm().GetUsernamePassword
                If Not LoggingSettings.LoginDetails.IsSet Then _
                   Throw New System.Configuration.ConfigurationErrorsException("Error getting login details")
            End If
        End Sub
        Private Shared ReadOnly Property FilePrefix(ByVal LogFolder As String) As String
            Get
                Return String.Format("{1}\{0:MMM-d yyyy HHmm-ss.FF}", Date.Now, LogFolder)
            End Get
        End Property
        Private Sub NewBadPagesTraceListener()
            AddListener(conBadPages, New BadPagesTraceListener(FilePrefix(LoggingSettings.Settings.LogFolder) & _
               " skipped.txt", LoggingSettings.Settings, LoggingSettings))
        End Sub
        Private Sub NewXHTMLTraceListener()
            AddListener(conXHTML, New XHTMLTraceListener(FilePrefix(LoggingSettings.Settings.LogFolder) & _
               " log.html", LoggingSettings))
        End Sub
        Private Sub NewWikiTraceListener()
            AddListener(conWiki, New WikiTraceListener(FilePrefix(LoggingSettings.Settings.LogFolder) & " log.txt", _
               LoggingSettings))
        End Sub
        Private ReadOnly Property FileNameFromActiveListener(ByVal Key As String) As String
            Get
                Return DirectCast(Listeners(Key), ITraceStatusProvider).TraceStatus.FileName
            End Get
        End Property
        Private Sub RemoveListenerAndReplaceWithSameType(ByVal Key As String)
            Dim str As String = FileNameFromActiveListener(Key)
            RemoveListener(Key)
            NewWikiTraceListener()
            Listeners(Key).WriteCommentAndNewLine("logging continued from " & str)
        End Sub
        Private Sub Busy()
            If Listeners.Count = 0 Then Exit Sub
            BusyCounter += 1
            LoggingSettings.Led1.Colour = Colour.Green
        End Sub
        Private Sub NotBusy()
            If Listeners.Count = 0 Then Exit Sub
            BusyCounter -= 1
            If BusyCounter = 0 AndAlso Not LoggingSettings.Led1.Colour = Colour.Blue _
               Then LoggingSettings.Led1.Colour = Colour.Red
        End Sub

        ' Overrides:
        Public Overrides Sub AddListener(ByVal Key As String, ByVal Listener As IMyTraceListener)
            MyBase.AddListener(Key, Listener)
            If Listener.Uploadable Then _
               AddHandler (DirectCast(Listener, TraceListenerUploadableBase)).Upload, AddressOf Me.Trace_Upload
        End Sub
        Public Overrides Sub RemoveListener(ByVal Key As String)
            Dim Listener As IMyTraceListener = Listeners(Key)

            If Listener.Uploadable Then _
               RemoveHandler (DirectCast(Listener, TraceListenerUploadableBase)).Upload, AddressOf Me.Trace_Upload

            MyBase.RemoveListener(Key)
        End Sub
        Public Overrides Sub Close()
            Busy()

            Dim upload As Boolean

            If (ContainsKey(conBadPages) OrElse ContainsKey(conWiki)) Then
                If LoggingSettings.Settings.UploadYN AndAlso MessageBox.Show("Upload logs?", _
            "Logging", MessageBoxButtons.YesNo, MessageBoxIcon.Question, MessageBoxDefaultButton.Button1) _
            = DialogResult.Yes Then Upload = True

                If Upload Then
                    If ContainsKey(conBadPages) Then DirectCast(Listeners(conBadPages), BadPagesTraceListener).UploadLog()
                    If ContainsKey(conWiki) Then DirectCast(Listeners(conWiki), WikiTraceListener).UploadLog()
                End If
            End If

            For Each t As KeyValuePair(Of String, IMyTraceListener) In Listeners
                t.Value.WriteCommentAndNewLine("closing all logs")
                If t.Value.Uploadable Then DirectCast(t.Value, TraceListenerUploadableBase).Close(Upload) _
                   Else t.Value.Close()
            Next
            Listeners.Clear()

            NotBusy()
        End Sub
#Region "Generic overrides"
        Public Overrides Sub ProcessingArticle(ByVal FullArticleTitle As String, ByVal NS As WikiFunctions.Namespaces)
            Busy()
            MyBase.ProcessingArticle(FullArticleTitle, NS)
            NotBusy()
        End Sub
        Public Overrides Sub SkippedArticle(ByVal SkippedBy As String, ByVal Reason As String)
            Busy()
            MyBase.SkippedArticle(SkippedBy, Reason)
            NotBusy()
        End Sub
        Public Overrides Sub SkippedArticleBadTag(ByVal SkippedBy As String, ByVal FullArticleTitle As String, ByVal NS As WikiFunctions.Namespaces)
            Busy()
            MyBase.SkippedArticleBadTag(SkippedBy, FullArticleTitle, NS)
            NotBusy()
        End Sub
        Public Overrides Sub SkippedArticleRedlink(ByVal FullArticleTitle As String, ByVal NS As WikiFunctions.Namespaces)
            Busy()
            MyBase.SkippedArticleRedlink(FullArticleTitle, NS)
            NotBusy()
        End Sub
        Public Overrides Sub Write(ByVal Text As String)
            Busy()
            MyBase.Write(Text)
            NotBusy()
        End Sub
        Public Overrides Sub WriteArticleActionLine(ByVal Line As String, ByVal PluginName As String)
            Busy()
            MyBase.WriteArticleActionLine(Line, PluginName)
            NotBusy()
        End Sub
        Public Overrides Sub WriteArticleActionLine1(ByVal Line As String, ByVal PluginName As String, ByVal VerboseOnly As Boolean)
            Busy()
            MyBase.WriteArticleActionLine1(Line, PluginName, VerboseOnly)
            NotBusy()
        End Sub
        Public Overrides Sub WriteBulletedLine(ByVal Line As String, ByVal Bold As Boolean, ByVal VerboseOnly As Boolean, Optional ByVal DateStamp As Boolean = False)
            Busy()
            MyBase.WriteBulletedLine(Line, Bold, VerboseOnly, DateStamp)
            NotBusy()
        End Sub
        Public Overrides Sub WriteComment(ByVal Line As String)
            Busy()
            MyBase.WriteComment(Line)
            NotBusy()
        End Sub
        Public Overrides Sub WriteCommentAndNewLine(ByVal Line As String)
            Busy()
            MyBase.WriteCommentAndNewLine(Line)
            NotBusy()
        End Sub
        Public Overrides Sub WriteLine(ByVal Line As String)
            Busy()
            MyBase.WriteLine(Line)
            NotBusy()
        End Sub
        Public Overrides Sub WriteTemplateAdded(ByVal Template As String, ByVal PluginName As String)
            Busy()
            MyBase.WriteTemplateAdded(Template, PluginName)
            NotBusy()
        End Sub
#End Region

        ' Callback from Settings control:
        Friend Sub PropertiesChange(ByVal JobNameHasChanged As Boolean)
            If LoggingSettings.Settings.LogFolder = LogFolder Then

                If LoggingSettings.Settings.LogXHTML Then
                    If Not ContainsKey(conXHTML) Then NewXHTMLTraceListener()
                ElseIf ContainsKey(conXHTML) Then
                    RemoveListener(conXHTML)
                End If

                If LoggingSettings.Settings.LogWiki Then
                    If ContainsKey(conWiki) Then
                        DirectCast(Listeners.Item(conWiki), WikiTraceListener).UploadLog(True)
                    Else
                        NewWikiTraceListener()
                    End If
                ElseIf ContainsKey(conWiki) Then
                    RemoveListener(conWiki)
                End If

                If LoggingSettings.Settings.LogBadPages Then
                    If ContainsKey(conBadPages) Then
                        DirectCast(Listeners.Item(conBadPages), BadPagesTraceListener).UploadLog(True)
                    Else
                        NewBadPagesTraceListener()
                    End If
                ElseIf ContainsKey(conBadPages) Then
                    RemoveListener(conBadPages)
                End If

            ElseIf HaveOpenFile Then ' folder has changed, close and reopen all active logs
                If ContainsKey(conWiki) Then RemoveListenerAndReplaceWithSameType(conWiki)
                If ContainsKey(conXHTML) Then RemoveListenerAndReplaceWithSameType(conXHTML)
                If ContainsKey(conBadPages) Then RemoveListenerAndReplaceWithSameType(conBadPages)
            End If

            CheckWeHaveLogInDetails()
        End Sub

        ' Trace listener child classes:
        ''' <summary>
        ''' Keeps track of logging statistics
        ''' </summary>
        ''' <remarks></remarks>
        Private NotInheritable Class TraceStatus
            Inherits Uploader.TraceStatus
            Private LinesLabel As Label, LinesSinceUploadLabel As Label, NumberOfUploadsLabel As Label
            Private Shared mUploadCount As Integer

            ' Initialisation
            Public Sub New(ByVal pLinesLabel As Label, ByVal pLinesSinceUploadLabel As Label, _
            ByVal pNumberOfUploadsLabel As Label, ByVal Uploadable As Boolean, ByVal FileN As String, _
            ByVal LogNameIs As String)
                MyBase.New(FileN, LogNameIs)
                LinesLabel = pLinesLabel
                LinesSinceUploadLabel = pLinesSinceUploadLabel
                If Not pNumberOfUploadsLabel Is Nothing Then
                    NumberOfUploadsLabel = pNumberOfUploadsLabel
                    NumberOfUploadsLabel.Text = mUploadCount.ToString
                End If
                pLinesLabel.Text = "0"
                If Uploadable Then pLinesSinceUploadLabel.Text = "0"
            End Sub

            ' Overrides:
            Public Overrides Sub Close()
                LinesLabel.Text = "N/A"
                LinesSinceUploadLabel.Text = "N/A"
            End Sub
            Public Overrides Property LinesWritten() As Integer
                Get
                    Return MyBase.LinesWritten
                End Get
                Set(ByVal value As Integer)
                    MyBase.LinesWritten = value
                    LinesLabel.Text = value.ToString
                End Set
            End Property
            Public Overrides Property LinesWrittenSinceLastUpload() As Integer
                Get
                    Return MyBase.LinesWrittenSinceLastUpload
                End Get
                Set(ByVal value As Integer)
                    MyBase.LinesWrittenSinceLastUpload = value
                    If Not LinesSinceUploadLabel Is Nothing Then LinesSinceUploadLabel.Text = value.ToString
                End Set
            End Property

            ' Extra:
            Public Property UploadsCount() As Integer
                Get
                    Return mUploadCount
                End Get
                Set(ByVal value As Integer)
                    mUploadCount = value
                    NumberOfUploadsLabel.Text = value.ToString
                End Set
            End Property
        End Class

        ''' <summary>
        ''' Logs in XHTML
        ''' </summary>
        ''' <remarks></remarks>
        Private NotInheritable Class XHTMLTraceListener
            Inherits Logging.XHTMLTraceListener
            Implements ITraceStatusProvider

            Private mTraceStatus As TraceStatus

            Public Sub New(ByVal FileName As String, ByVal LS As PluginLogging)
                MyBase.New(FileName, LS.Settings.LogVerbose)
                mTraceStatus = New TraceStatus(LS.XHTMLLinesLabel, Nothing, Nothing, False, FileName, conXHTML)
            End Sub
            Public ReadOnly Property TraceStatus() As WikiFunctions.Logging.Uploader.TraceStatus _
            Implements ITraceStatusProvider.TraceStatus
                Get
                    Return mTraceStatus
                End Get
            End Property
        End Class

        ''' <summary>
        ''' Logs in wiki format
        ''' </summary>
        ''' <remarks></remarks>
        Private NotInheritable Class WikiTraceListener
            Inherits Logging.WikiTraceListener

            Public Sub New(ByVal FileName As String, ByVal LS As PluginLogging)
                MyBase.New(LS.Settings, _
                   New TraceStatus(LS.WikiLinesLabel, LS.WikiLinesSinceUploadLabel, LS.UploadsCountLabel, _
                      LS.Settings.UploadYN, FileName, conWiki))
            End Sub
        End Class

        ''' <summary>
        ''' Logs pages which have bad templates
        ''' </summary>
        ''' <remarks></remarks>
        Private NotInheritable Class BadPagesTraceListener
            Inherits TraceListenerUploadableBase

            Public Sub New(ByVal FileName As String, ByVal UploadSettings As UploadableLogSettings2, ByVal LS As PluginLogging)
                MyBase.New(UploadSettings, _
                   New TraceStatus(LS.BadTagsLinesLabel, LS.BadTagsLinesSinceUploadLabel, LS.UploadsCountLabel, _
                   LS.Settings.UploadYN, FileName, conBadPages))
                WriteBulletedLine("Logging: [[User:Kingbotk/Plugin/WikiFunctions2|WikiFunctions2]].dll v" & _
                   WikiFunctions2.Version.ToString, False, False)
            End Sub

            ' Overrides:
            Public Overrides Sub SkippedArticleBadTag(ByVal SkippedBy As String, ByVal FullArticleTitle As String, _
            ByVal NS As Namespaces)
                MyBase.WriteLine(GetArticleTemplate(FullArticleTitle, NS) & " (skipped by the " & SkippedBy _
                   & " plugin; bad tag)")
            End Sub
            Public Overrides Sub SkippedArticleRedlink(ByVal FullArticleTitle As String, ByVal NS As Namespaces)
                MyBase.WriteLine(GetArticleTemplate(FullArticleTitle, NS) & " (skipped by the Plugin Manager; " & _
                   "attached article doesn't exist - maybe deleted?)")
            End Sub
            Public Overrides Sub WriteLine(ByVal value As String, Optional ByVal CheckCounter As Boolean = True)
                MyBase.WriteLine(value, CheckCounter)
            End Sub
            Public Overrides Sub WriteComment(ByVal Line As String)
                MyBase.Write("<!-- " & Line & " -->")
            End Sub
            Public Overrides Sub WriteCommentAndNewLine(ByVal Line As String)
                MyBase.WriteLine("<!-- " & Line & " -->")
            End Sub

            ' Overrides - do nothing:
            Public Overrides Sub WriteBulletedLine(ByVal Line As String, ByVal bold As Boolean, _
               ByVal b As Boolean, Optional ByVal DateStamp As Boolean = False)
            End Sub
            Public Overrides Sub ProcessingArticle(ByVal ArticleFullTitle As String, ByVal NS As Namespaces)
            End Sub
            Public Overrides Sub WriteArticleActionLine(ByVal Line As String, ByVal PluginName As String)
            End Sub
            Public Overrides Sub WriteTemplateAdded(ByVal Template As String, ByVal PluginName As String)
            End Sub
            Public Overrides Sub SkippedArticle(ByVal SkippedBy As String, ByVal Reason As String)
            End Sub
        End Class

        Public Sub New(ByVal pActivePlugins As List(Of PluginBase))
            MyBase.new()
            ActivePlugins = pActivePlugins
        End Sub
        Friend WriteOnly Property LS() As PluginLogging
            Set(ByVal value As PluginLogging)
                LoggingSettings = value
            End Set
        End Property
    End Class
End Namespace

File: AWB/AWB/MyTrace.cs
/*
(C) 2007 Stephen Kennedy (Kingboyk) http://www.sdk-software.com/

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

// From the Kingbotk plugin. Converted from VB to C#

using System;
using System.Collections.Generic;
using System.Text;
using WikiFunctions;
using WikiFunctions.Logging;
using WikiFunctions.Plugin;
using System.Windows.Forms;
using WikiFunctions.Logging.Uploader;

namespace AutoWikiBrowser.Logging
{
    /// <summary>
    /// Logging manager
    /// </summary>
    internal sealed class MyTrace : TraceManager, IAWBTraceListener
    {
        //private new Dictionary<string, IAWBTraceListener> Listeners = new Dictionary<string, IAWBTraceListener>();

        private LoggingSettings LoggingSettings;
        private static string LogFolder = "";
        private int BusyCounter = 0;
        private bool mIsUploading = false;

        private const string conWiki = "Wiki";
        private const string conXHTML = "XHTML";

        // The most important stuff:
        internal void Initialise()
        {
            try
            {
                LoggingSettings.Initialised = true;
                LogFolder = LoggingSettings.Settings.LogFolder;

                if (LoggingSettings.Settings.LogXHTML || LoggingSettings.Settings.LogWiki)
                {
                    if (!(System.IO.Directory.Exists(LoggingSettings.Settings.LogFolder)))
                    {
                        LogFolder = Application.StartupPath;
                    }

                    if (LoggingSettings.Settings.LogXHTML)
                    {
                        NewXHTMLTraceListener();
                    }
                    if (LoggingSettings.Settings.LogWiki)
                    {
                        NewWikiTraceListener();
                    }
                }
                foreach (KeyValuePair<string, IMyTraceListener> t in Listeners)
                {
                    ((IAWBTraceListener)t.Value).WriteBulletedLine(Variables.LoggingStartButtonClicked, true, false, true);
                }
                CheckWeHaveLogInDetails();
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }
        }

        private void TraceUploadEventHandler(TraceListenerUploadableBase Sender, ref bool Success)
        {
            Success = base.UploadHandler(Sender, LoggingSettings.Settings.LogTitle, 
                LoggingSettings.Settings.WikifiedCategory, LoggingSettings.Settings.GlobbedUploadLocation + "/" + 
                Sender.PageName.Replace(LoggingSettings.Props.conUploadCategoryIsJobName, 
                LoggingSettings.Settings.Category), LoggingSettings.Settings.LinksToLog(), 
                LoggingSettings.Settings.UploadOpenInBrowser, LoggingSettings.Settings.UploadAddToWatchlist, 
                LoggingSettings.Props.UserName, Variables.AWBVersionString(GlobalObjects.AWB.AWBVersionString) +
                Plugins.Plugin.GetPluginsWikiTextBlock(), Variables.AWBLoggingEditSummary +
                Variables.UploadingLogDefaultEditSummary, Variables.AWBLoggingEditSummary +
                Variables.UploadingLogEntryDefaultEditSummary, GlobalObjects.AWB, LoggingSettings.LoginDetails);

            if (Success)
            {
                ((TraceStatus)Sender.TraceStatus).UploadsCount += 1;
            }
        }
        protected override bool StartingUpload(TraceListenerUploadableBase Sender)
        {
            if (!(Sender.TraceStatus.LogName == conWiki))
                return false;

            mIsUploading = true;
            LoggingSettings.LEDColour = WikiFunctions.Controls.Colour.Blue;
            Application.DoEvents();
            return true;
        }

        protected override void FinishedUpload()
        {
            if (BusyCounter == 0)
                LoggingSettings.LEDColour = WikiFunctions.Controls.Colour.Red;
            else
                LoggingSettings.LEDColour = WikiFunctions.Controls.Colour.Green;

            mIsUploading = false;
        }

        // State:
        internal bool HaveOpenFile
        {
            get { return Listeners.Count > 0; }
        }

        internal bool IsUploading
        {
            get { return mIsUploading; }
        }

        // Private:
        private void CheckWeHaveLogInDetails()
        {
            if (this.Uploadable && !LoggingSettings.LoginDetails.IsSet)
            {
                // TODO: This needs to be imported from plugin and fixed, but what about AWBProfiles? Encyryption? etc
                //LoggingSettings.LoginDetails = new LoginForm().GetUsernamePassword;
                //if (!LoggingSettings.LoginDetails.IsSet)
                //{
                //    throw new System.Configuration.ConfigurationErrorsException("Error getting login details");
                //}
            }
        }
        private static string GetFilePrefix(string LogFolder)
        {
            return string.Format("{1}\\{0:MMM-d yyyy HHmm-ss.FF}", System.DateTime.Now, LogFolder);
        }
        private void NewXHTMLTraceListener()
        {
            AddListener(conXHTML, new XHTMLTraceListener(GetFilePrefix(LoggingSettings.Settings.LogFolder) + 
                " log.html", LoggingSettings));
        }
        private void NewWikiTraceListener()
        {
            AddListener(conWiki, new WikiTraceListener(GetFilePrefix(LoggingSettings.Settings.LogFolder) + 
                " log.txt", LoggingSettings));
        }
        private string GetFileNameFromActiveListener(string Key)
        {
            return ((ITraceStatusProvider)(Listeners[Key])).TraceStatus.FileName;
        }
        private void RemoveListenerAndReplaceWithSameType(string Key)
        {
            string str = GetFileNameFromActiveListener(Key);
            RemoveListener(Key);
            NewWikiTraceListener();
            Listeners[Key].WriteCommentAndNewLine("logging continued from " + str);
        }
        private void Busy()
        {
            if (Listeners.Count == 0)
                return;
            BusyCounter += 1;
            LoggingSettings.LEDColour = WikiFunctions.Controls.Colour.Green;
        }
        private void NotBusy()
        {
            if (Listeners.Count == 0)
                return;
            BusyCounter -= 1;
            if (BusyCounter == 0 && !(LoggingSettings.LEDColour == WikiFunctions.Controls.Colour.Blue))
            {
                LoggingSettings.LEDColour = WikiFunctions.Controls.Colour.Red;
            }
        }

        private bool BadPagesLogToUpload
        {
            get { return false; }
        }
        private bool WikiLogToUpload
        {
            get { return (ContainsKey(conWiki) && ((WikiTraceListener)(Listeners[conWiki])).TraceStatus.LinesWrittenSinceLastUpload > 1); }
        }

        // Overrides:
        public override void AddListener(string Key, IMyTraceListener Listener)
        {
            if (base.ContainsKey(Key))
            {
                base.RemoveListener(Key);
            }

            base.AddListener(Key, Listener);
            if (Listener.Uploadable)
            {
                ((TraceListenerUploadableBase)Listener).Upload += 
                    new WikiFunctions.Logging.TraceListenerUploadableBase.UploadEventHandler(
                    this.TraceUploadEventHandler);
            }
        }
        public override void RemoveListener(string Key)
        {
            if (!Listeners.ContainsKey(Key)) return;
            IMyTraceListener Listener = Listeners[Key];

            if (Listener.Uploadable)
            {
                ((TraceListenerUploadableBase)Listener).Upload -= 
                    new WikiFunctions.Logging.TraceListenerUploadableBase.UploadEventHandler(
                    this.TraceUploadEventHandler);
            }

            base.RemoveListener(Key);
        }
        public override void Close()
        {
            Busy();

            bool upload = false;

            if (LoggingSettings.Settings.UploadYN && (BadPagesLogToUpload || WikiLogToUpload) && MessageBox.Show("Upload logs?", "Logging", MessageBoxButtons.YesNo, MessageBoxIcon.Question, MessageBoxDefaultButton.Button1) == DialogResult.Yes)
                upload = true;

            foreach (KeyValuePair<string, IMyTraceListener> t in Listeners)
            {
                ((IAWBTraceListener)t.Value).WriteCommentAndNewLine("closing all logs");
                if (t.Value.Uploadable)
                {
                    ((TraceListenerUploadableBase)t.Value).Close(upload);
                }
                else
                {
                    t.Value.Close();
                }
            }
            Listeners.Clear();

            NotBusy();
        }

        // Friend:
        internal void UploadWikiLog()
        {
            if (ContainsKey(conWiki))
            {
                ((WikiTraceListener)(Listeners[conWiki])).UploadLog();
            }
        }

        #region Generic overrides
        public void AWBSkipped(string Reason)
        {
            Busy();
            foreach (KeyValuePair<string, IMyTraceListener> Listener in Listeners)
            {
                ((IAWBTraceListener)Listener.Value).AWBSkipped(Reason);
            }
        }
        public void PluginSkipped()
        {
            Busy();
            foreach (KeyValuePair<string, IMyTraceListener> Listener in Listeners)
            {
                ((IAWBTraceListener)Listener.Value).PluginSkipped();
            }
        }
        public void UserSkipped()
        {
            Busy();
            foreach (KeyValuePair<string, IMyTraceListener> Listener in Listeners)
            {
                ((IAWBTraceListener)Listener.Value).UserSkipped();
            }
        }
        public override void ProcessingArticle(string FullArticleTitle, WikiFunctions.Namespaces NS)
        {
            Busy();
            base.ProcessingArticle(FullArticleTitle, NS);
            NotBusy();
        }
        public override void SkippedArticle(string SkippedBy, string Reason)
        {
            Busy();
            base.SkippedArticle(SkippedBy, Reason);
            NotBusy();
        }
        public override void SkippedArticleBadTag(string SkippedBy, string FullArticleTitle, WikiFunctions.Namespaces NS)
        {
            Busy();
            base.SkippedArticleBadTag(SkippedBy, FullArticleTitle, NS);
            NotBusy();
        }
        public override void SkippedArticleRedlink(string SkippedBy, string FullArticleTitle, WikiFunctions.Namespaces NS)
        {
            Busy();
            base.SkippedArticleRedlink(SkippedBy, FullArticleTitle, NS);
            NotBusy();
        }
        public override void Write(string Text)
        {
            Busy();
            base.Write(Text);
            NotBusy();
        }
        public override void WriteArticleActionLine(string Line, string PluginName)
        {
            Busy();
            base.WriteArticleActionLine(Line, PluginName);
            NotBusy();
        }
        public override void WriteArticleActionLine1(string Line, string PluginName, bool VerboseOnly)
        {
            Busy();
            base.WriteArticleActionLine1(Line, PluginName, VerboseOnly);
            NotBusy();
        }
        public override void WriteBulletedLine(string Line, bool Bold, bool VerboseOnly, bool DateStamp)
        {
            Busy();
            base.WriteBulletedLine(Line, Bold, VerboseOnly, DateStamp);
            NotBusy();
        }
        public override void WriteComment(string Line)
        {
            Busy();
            base.WriteComment(Line);
            NotBusy();
        }
        public override void WriteCommentAndNewLine(string Line)
        {
            Busy();
            base.WriteCommentAndNewLine(Line);
            NotBusy();
        }
        public override void WriteLine(string Line)
        {
            Busy();
            base.WriteLine(Line);
            NotBusy();
        }
        public override void WriteTemplateAdded(string Template, string PluginName)
        {
            Busy();
            base.WriteTemplateAdded(Template, PluginName);
            NotBusy();
        }
        #endregion

        // Callback from Settings control:
        internal void PropertiesChange(bool JobNameHasChanged)
        {
            if (LoggingSettings.Settings.LogFolder == LogFolder)
            {
                if (LoggingSettings.Settings.LogXHTML)
                    if (!(ContainsKey(conXHTML)))
                        NewXHTMLTraceListener();
                else if (ContainsKey(conXHTML))
                    RemoveListener(conXHTML);

                if (LoggingSettings.Settings.LogWiki)
                    if (!(ContainsKey(conWiki)))
                        NewWikiTraceListener();
                else if (ContainsKey(conWiki))
                    RemoveListener(conWiki);
            }
            else if (HaveOpenFile) // folder has changed, close and reopen all active logs
            {
                if (ContainsKey(conWiki))
                    RemoveListenerAndReplaceWithSameType(conWiki);
                if (ContainsKey(conXHTML))
                    RemoveListenerAndReplaceWithSameType(conXHTML);
            }

            CheckWeHaveLogInDetails();
        }

        // Trace listener child classes:
        /// <summary>
        /// Keeps track of logging statistics
        /// </summary>
        /// <remarks></remarks>
        private sealed class TraceStatus : WikiFunctions.Logging.Uploader.TraceStatus
        {
            private Label LinesLabel;
            private Label LinesSinceUploadLabel;
            private Label NumberOfUploadsLabel;
            private static int mUploadCount;

            // Initialisation
            public TraceStatus(Label pLinesLabel, Label pLinesSinceUploadLabel, Label pNumberOfUploadsLabel, 
                bool Uploadable, string FileN, string LogNameIs) : base(FileN, LogNameIs)
            {
                LinesLabel = pLinesLabel;
                LinesSinceUploadLabel = pLinesSinceUploadLabel;
                if (pNumberOfUploadsLabel != null)
                {
                    NumberOfUploadsLabel = pNumberOfUploadsLabel;
                    NumberOfUploadsLabel.Text = mUploadCount.ToString();
                }
                pLinesLabel.Text = "0";
                if (Uploadable)
                {
                    pLinesSinceUploadLabel.Text = "0";
                }
            }

            // Overrides:
            public override void Close()
            {
                LinesLabel.Text = "N/A";
                LinesSinceUploadLabel.Text = "N/A";
            }
            public override int LinesWritten
            {
                get { return base.LinesWritten; }
                set
                {
                    base.LinesWritten = value;
                    LinesLabel.Text = value.ToString();
                }
            }
            public override int LinesWrittenSinceLastUpload
            {
                get { return base.LinesWrittenSinceLastUpload; }
                set
                {
                    base.LinesWrittenSinceLastUpload = value;
                    if (LinesSinceUploadLabel != null)
                    {
                        LinesSinceUploadLabel.Text = value.ToString();
                    }
                }
            }

            // Extra:
            public int UploadsCount
            {
                get { return mUploadCount; }
                set
                {
                    mUploadCount = value;
                    NumberOfUploadsLabel.Text = value.ToString();
                }
            }
        }

        /// <summary>
        /// Logs in XHTML
        /// </summary>
        /// <remarks></remarks>
        private sealed class XHTMLTraceListener : WikiFunctions.Logging.XHTMLTraceListener,
            WikiFunctions.Logging.Uploader.ITraceStatusProvider, IAWBTraceListener
        {
            private TraceStatus mTraceStatus;

            public XHTMLTraceListener(string FileName, LoggingSettings LS)
                : base(FileName, LS.Settings.LogVerbose)
            {
                mTraceStatus = new TraceStatus(LS.XHTMLLinesLabel, null, null, false, FileName, conXHTML);
            }
            public WikiFunctions.Logging.Uploader.TraceStatus TraceStatus
            {
                get { return mTraceStatus; }
            }
            public void AWBSkipped(string Reason)
            {
                this.SkippedArticle("AWB", Reason);
            }
            public void UserSkipped()
            { this.SkippedArticle(Variables.StringUser, Variables.StringUserSkipped); }
            public void PluginSkipped()
            { this.SkippedArticle(Variables.StringPlugin, Variables.StringPluginSkipped); }
        }

        /// <summary>
        /// Logs in wiki format
        /// </summary>
        /// <remarks></remarks>
        private sealed class WikiTraceListener : WikiFunctions.Logging.WikiTraceListener, IAWBTraceListener
        {
            public WikiTraceListener(string FileName, LoggingSettings LS)
                : base(LS.Settings, new TraceStatus(LS.WikiLinesLabel, LS.WikiLinesSinceUploadLabel, 
                LS.UploadsCountLabel, LS.Settings.UploadYN, FileName, conWiki)) { }

            public void AWBSkipped(string Reason)
            {
                this.SkippedArticle("AWB", Reason);
            }
            public void UserSkipped()
            { this.SkippedArticle(Variables.StringUser, Variables.StringUserSkipped); }
            public void PluginSkipped()
            { this.SkippedArticle(Variables.StringPlugin, Variables.StringPluginSkipped); }
        }

        internal LoggingSettings LS
        {
            get { return LoggingSettings; }
            set { LoggingSettings = value; }
        }

        protected override string ApplicationName
        {
            get { return "AutoWikiBrowser logging manager"; }
        }

        // Allow plugins to turn off all logging:
        public void TurnOffLogging()
        {
            Close(); LS.TurnOffLogging();
        }

        internal static AWBLogListener InitialiseLogListener(ArticleEx Article)
        { return null;/* Article.InitialiseLogListener("AWB", GlobalObjects.MyTrace);*/ }
        // TODO: At some point we need to *remove* the listener for the article ("AWB")
        // Plugin did it at the end of ProcessArticle(). We also do it, but a bit late, in AddListener().
    }
}